File: Microsoft/Largest Divisible Subset.py
class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        nums.sort()
        n=len(nums)
        dp=[1 for i in range(n)]
        hash=[i for i in range(n)]
        for i in range(n):
            for j in range(i):
                if nums[i]%nums[j]==0 and dp[i]<1+dp[j]:
                    dp[i]=1+dp[j]
                    hash[i]=j
        maxi=1
        max_ind=0
        for i in range(n):
            if dp[i]>maxi:
                maxi=max(maxi,dp[i])
                max_ind=i
        temp=[]
        temp.append(nums[max_ind])
        while hash[max_ind]!=max_ind:
            max_ind=hash[max_ind]
            temp.append(nums[max_ind])
        temp.reverse()
        return temp

File: Microsoft/Largest_Divisible_Subset.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    vector<int> largestDivisibleSubset(vector<int> &nums)
    {
        sort(nums.begin(), nums.end());

        int n = nums.size();
        vector<int> dp(n, 1);
        vector<int> prevInd(n);
        int maxi = 1;
        int lastInd = 0;
        for (int i = 0; i < n; i++)
        {
            prevInd[i] = i;
            for (int prev = 0; prev < i; prev++)
            {
                if (nums[i] % nums[prev] == 0 && 1 + dp[prev] > dp[i])
                {
                    dp[i] = 1 + dp[prev];
                    prevInd[i] = prev;
                }
            }

            if (dp[i] > maxi)
            {
                maxi = dp[i];
                lastInd = i;
            }
        }

        vector<int> ans;
        ans.push_back(nums[lastInd]);
        while (prevInd[lastInd] != lastInd)
        {
            lastInd = prevInd[lastInd];
            ans.push_back(nums[lastInd]);
        }

        return ans;
    }
};

int main()
{
    Solution obj;
    cout << "Constraints:\n1 <= nums.length <= 1000\n1 <=nums[i] <= 2 * 109\nAll the integers in nums are unique." << endl;

    int num;
    cout << "Enter Number of element in array:-";
    cin >> num;
    vector<int> vec(num);
    for (int i = 0; i < num; i++)
    {
        cin >> vec[i];
    }

    vector<int> lds = obj.largestDivisibleSubset(vec);

    for (auto a : lds)
        cout << a << " ";
    cout << endl;
}